internal/page/donor/witnessing_as_independent_witness_test.go

```diff
@@ -0,0 +1,395 @@
+		On("Execute", w, &witnessingAsIndependentWitnessData{
+			App:  testAppData,
+			Lpa:  &page.Lpa{},
+			Form: &witnessingAsIndependentWitnessForm{},
+		}).
+		Return(expectedError)
+
+	err := WitnessingAsIndependentWitness(template.Execute, nil, time.Now)(testAppData, w, r, &page.Lpa{})
+	resp := w.Result()
+
+	assert.Equal(t, expectedError, err)
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+}
+
+func TestPostWitnessingAsIndependentWitness(t *testing.T) {
+	form := url.Values{
+		"witness-code": {"1234"},
+	}
+
+	w := httptest.NewRecorder()
+	r, _ := http.NewRequest(http.MethodPost, "/", strings.NewReader(form.Encode()))
+	r.Header.Add("Content-Type", page.FormUrlEncoded)
+	now := time.Now()
+
+	donorStore := newMockDonorStore(t)
+	donorStore.
+		On("Put", r.Context(), &page.Lpa{
+			ID:                              "lpa-id",
+			DonorIdentityUserData:           identity.UserData{OK: true, Provider: identity.OneLogin},
+			IndependentWitnessCodes:         page.WitnessCodes{{Code: "1234", Created: now}},
+			WitnessedByIndependentWitnessAt: now,
+		}).
+		Return(nil)
+
+	err := WitnessingAsIndependentWitness(nil, donorStore, func() time.Time { return now })(testAppData, w, r, &page.Lpa{
+		ID:                      "lpa-id",
+		DonorIdentityUserData:   identity.UserData{OK: true, Provider: identity.OneLogin},
+		IndependentWitnessCodes: page.WitnessCodes{{Code: "1234", Created: now}},
+	})
+	resp := w.Result()
+
+	assert.Nil(t, err)
+	assert.Equal(t, http.StatusFound, resp.StatusCode)
+	assert.Equal(t, page.Paths.WitnessingAsCertificateProvider.Format("lpa-id"), resp.Header.Get("Location"))
+}
+
+func TestPostWitnessingAsIndependentWitnessWhenIdentityConfirmed(t *testing.T) {
+	form := url.Values{
+		"witness-code": {"1234"},
+	}
+
+	w := httptest.NewRecorder()
+	r, _ := http.NewRequest(http.MethodPost, "/", strings.NewReader(form.Encode()))
+	r.Header.Add("Content-Type", page.FormUrlEncoded)
+	now := time.Now()
+
+	lpa := &page.Lpa{
+		ID:                              "lpa-id",
+		DonorIdentityUserData:           identity.UserData{OK: true, Provider: identity.OneLogin},
+		IndependentWitness:              actor.IndependentWitness{},
+		IndependentWitnessCodes:         page.WitnessCodes{{Code: "1234", Created: now}},
+		WitnessedByIndependentWitnessAt: now,
+	}
+	donorStore := newMockDonorStore(t)
+	donorStore.
+		On("Put", r.Context(), lpa).
+		Return(nil)
+
+	err := WitnessingAsIndependentWitness(nil, donorStore, func() time.Time { return now })(testAppData, w, r, &page.Lpa{
+		ID:                      "lpa-id",
+		DonorIdentityUserData:   identity.UserData{OK: true, Provider: identity.OneLogin},
+		IndependentWitness:      actor.IndependentWitness{},
+		IndependentWitnessCodes: page.WitnessCodes{{Code: "1234", Created: now}},
+	})
+	resp := w.Result()
+
+	assert.Nil(t, err)
+	assert.Equal(t, http.StatusFound, resp.StatusCode)
+	assert.Equal(t, page.Paths.WitnessingAsCertificateProvider.Format("lpa-id"), resp.Header.Get("Location"))
+}
```

Comment:
Is there anything different being asserted here besides a witness being set on the LPA passed to the handler?

Reply:
you're right, it was some bad copypasta